Kafka 4507: The client should send older versions of requests to the broker if necessary

cmccabe opened the pull request:
KAFKA-4507

The client should send older versions of requests to the broker if necessary.

Note: This builds on top of KAFKA-3600, which has not yet been committed yet.

cmccabe commented:
Previously, there were many places in the code that were creating AbstractRequest objects directly and passing them to the NetworkClient. Unfortunately, AbstractRequest objects are version-specific.  For example, a ListOffsetRequest object of version 0 will have different fields inside its Struct than a ListOffsetRequest object of version 1.  As soon as the message constructor is invoked, the contained Structs are formatted in the version-specific format.

This change adds the AbstractRequest#Builder types.  Client code in the Fetcher and Producer, and other places, can create these objects and pass them to the NetworkClient.  The NetworkClient can then decide what version of each request  to use based on the ApiVersionsRequest data.  When the version that needs to be used is too old to support a necessary feature, an error is returned to the     upper layer, which can then decide what to do.

ijuma commented:
@cmccabe, can you please rebase this branch? Thanks!

cmccabe commented:
Rebased on trunk and fixed some unit tests

ijuma commented:
Sorry, can you please rebase again?

cmccabe commented:
Rebased

cmccabe committed:
KAFKA-4507: The client should send older versions of requests to the broker if necessary

cmccabe commented:
Also removed a few places where this change duplicated KAFKA-4548, and reverted a change to the test log4j.properties file

cmccabe commented:
Fix AuthorizerIntegrationTest, FetchRequestTest, SaslApiVersionsRequestTest, PlaintextConsumerTest.

Fixed a bug where Fetcher#beginningOffsets and Fetcher#endOffsets incorrectly required a broker with the new ListOffsetsRequest RPC.  Will add this to CompatibilityTest.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/ClientResponse.java:
```diff
@@ -62,6 +67,10 @@ public boolean wasDisconnected() {
         return disconnected;
     }
 
+    public boolean versionMismatch() {
```
It seems like this field is not used anywhere.

cmccabe replied:
This should be used any time that there is a version mismatch problem between the message we are trying to send and the message versions we can actually send to the desired broker.  I fixed the Fetcher to use this instead of doing its own thing to detect version problems.  I also change the type of this to RuntimeException, so that the exact exception which the builder raised can be preserved here.

cmccabe committed:
Add NodeApiVersions.java and unit tests

(cherry picked from commit a4e2bdbeaf832d55985a056beea769805703f2dd)

cmccabe committed:
Improve NodeApiVersionsTest

(cherry picked from commit 51fc001b938c2e0318fb4f44290ad3c243e7fbaa)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -349,19 +347,20 @@ private void sendProduceRequest(long now, int destination, short acks, int timeo
-        ProduceRequest produceRequest = new ProduceRequest(acks, timeout, produceRecordsByPartition);
-        RequestHeader header = this.client.nextRequestHeader(ApiKeys.PRODUCE);
+        ProduceRequest.Builder requestBuilder =
+                new ProduceRequest.Builder(acks, timeout, produceRecordsByPartition);
```
I think we need to enforce a minimum version of 2 since older versions would expect an older version of the message format. This is also an easy way to detect that we're talking with a version of the broker older than 0.10. I'm wondering if it would make sense to raise an exception to the user immediately when we connect to such a broker instead of waiting until we send an incompatible request?

cmccabe replied:
Sure, we can enforce a minimum version of 2 here.  Brokers earlier than 0.10 will automatically disconnect our clients, since those brokers don't support ApiVersionsRequest, which is the first thing we send after KAFKA-3600.

ijuma replied:
Unfortunately, 0.9.0.1 has a bug where it won't disconnect when it receives an unknown request. It will keep the connection open until the connection reaper comes along. A subsequent request can cause an immediate disconnection though.

cmccabe replied:
Hmm, how would the 0.9.0.1 broker even know that it was getting another request?  it doesn't know the size of the first request so it doesn't know where that request ends and a new one begins.

cmccabe replied:
It sounds like we would be in the clear then, since we're sending an ApiVersionsRequest followed by a follow-on request?

ijuma replied:
The broker assumes that all requests are consistent with regards to the common request header elements. If we set the minimum versions correctly, we should be in the clear, I think. It would be good to have a test for 0.9.0.1, but we can do that in a follow-up.

cmccabe committed:
Fetcher.java: fix indentation

cmccabe committed:
ProduceRequest: disable creating versions earlier than 2

cmccabe committed:
Add builders for CreateTopicsRequest and DeleteTopicsRequest, so that they can be used with send() etc.

cmccabe committed:
Fix checkStyle

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java:
```diff
@@ -13,45 +13,98 @@
+            this.topics = topics;
+        }
+
+        public List<String> getTopics() {
```
Nit: in Kafka, we don't use `get` prefix for accessors.

cmccabe replied:
ok, I'll remove that here (and elsewhere)

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/ListOffsetResponse.java:
```diff
@@ -89,6 +90,31 @@ public PartitionData(short errorCode, long timestamp, long offset) {
+            bld.append("PartitionData{").
+                append("errorCode: ").append((int) errorCode);
+            if (timestamp == null) {
+                bld.append(", timestamp: null");
```
This is not needed, `StringBuilder` appends `null` for you.

cmccabe replied:
k

ijuma reviewed (commented):
Thanks for the PR. I have more comments that I will post in a bit, but in the meantime, I wanted to ask: are we handling the case where pre-KIP-74 brokers return a RecordTooLargeException?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/ClientRequest.java:
```diff
@@ -21,40 +22,46 @@
-            ", header=" + header +
-            ", body=" + body +
+            ", destination=" + destination +
+            ", requestBuilder=" + requestBuilder +
```
Nit: it may make sense to move the requestBuilder to the end as it's likely to be bigger and make it harder to read the other fields.

cmccabe replied:
yeah

cmccabe replied:
yeah

ijuma commented on clients/src/main/java/org/apache/kafka/clients/ClientRequest.java:
```diff
@@ -83,4 +95,11 @@ public long createdTimeMs() {
+        return correlationId;
+    }
+
+    public AbstractRequest.Builder getRequestBuilder() {
```
We have a method `requestBuilder` that is the same as this, so we can remove this one.

cmccabe committed:
Accessors should not start with 'get'

cmccabe committed:
Don't special-case null when using the stringbuilder.  remove unused variable

cmccabe committed:
Some fixes to the system which handles mismatches between client versions and broker versions

If a message cannot be sent to a broker, because it is too old, an
exception will be thrown from AbstractRequest#Builder#build.  This
exception will be picked up in NetworkClient and used to create a
ClientResponse.  We now include the whole exception in the
ClientResponse, for extra detail and clarity in case it is needed.

In the Fetcher, we don't need to check for negative timestamps when
requireTimestamps is set, since this will be caught by the builder's
build method.  This check also was not reliable since sometimes no
offsets at all would be returned, and it would appear as if the broker
supported timestamps, when it did not.

cmccabe commented:
Good point about KIP-74 compatibility handling.  it sounds like we will have to reproduce the pre-KIP-74 behavior of "getting stuck" when messages are too big rather than the new behavior of returning at least one message.  I'll add back in that code shortly and work on a compatibility test for it.  The other comments should be addressed now in the latest patch.

cmccabe committed:
Fix checkstyle

ijuma commented on clients/src/main/java/org/apache/kafka/clients/KafkaClient.java:
```diff
@@ -112,24 +111,21 @@
-     */
-    public void wakeup();
-
+    public ClientRequest newClientRequest(String nodeId, AbstractRequest.Builder requestBuilder,
```
Nit: should this be `newRequest` since it's a method in `KafkaClient`? Or is it clearer how it is?

cmccabe replied:
Hmm.  I think if it were just newRequest, I would wonder whether it was a new AbstractRequest or a new ClientRequest.  What do you think?

ijuma replied:
I was debating that myself before writing the comment. Let's leave it as is then.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1351,7 +1348,9 @@ public OffsetAndMetadata committed(TopicPartition partition) {
-            Map<String, List<PartitionInfo>> topicMetadata = fetcher.getTopicMetadata(new MetadataRequest(Collections.singletonList(topic)), requestTimeoutMs);
+            Map<String, List<PartitionInfo>> topicMetadata =
+                    fetcher.getTopicMetadata(new MetadataRequest.Builder(
+                            Collections.singletonList(topic)), requestTimeoutMs);
```
Nit: do we really need 3 lines for this? Seems like we could do it in two. In Kafka, we generally go for wider lines than the traditional 80 columns (100 to 120 is common).

cmccabe replied:
k

cmccabe replied:
ok

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndTimestamp.java:
```diff
@@ -21,7 +21,8 @@
  *
- * Both offset and timestamp are non-negative.
+ * offset must be non-negative.
+ * If timestamp is negative, this object does not contain a valid timestamp.
```
Can we expand on when this can happen?

cmccabe replied:
I added a comment

cmccabe replied:
I added a comment

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/KafkaClient.java:
```diff
@@ -112,24 +111,21 @@
-    public void wakeup();
-
+    public ClientRequest newClientRequest(String nodeId, AbstractRequest.Builder requestBuilder,
+                                   long createdTimeMs, boolean expectResponse,
```
nit: this alignment is weird. Are you using intellij defaults?

cmccabe replied:
ok, I realigned it with the left paren

cmccabe replied:
fixed

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -296,31 +253,76 @@ public void send(ClientRequest request, long now) {
+            // that we can send data.  If the request is internal, we trust
+            // that that internal code has done this validation.  Validation
+            // will be slightly different for some internal requests (for
+            // example, ApiVersionsRequests can be sent prior to being in
```
Does it make sense to special case the ApiVersion request only instead of skipping the check for all internal requests?

cmccabe replied:
Hmm.  I haven't thought about it that hard, but I think special casing will probably get messy here.  It also duplicates work done elsewhere to check if the message is sendable.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -296,31 +253,76 @@ public void send(ClientRequest request, long now) {
+            // This would be the case when sending the initial ApiVersionRequest which
+            // fetches the version information itself, for example.
+            if (versionInfo == null) {
+                if (log.isTraceEnabled()) {
```
nit: we usually skip braces for one-line branches like this. A few more below.

cmccabe replied:
k

cmccabe replied:
k

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -296,31 +253,76 @@ public void send(ClientRequest request, long now) {
+                    log.trace("When sending message of type {} to node {}, the best usable " +
+                            "version is {}", clientRequest.apiKey(), nodeId, version);
+                }
+                if (version < 0) {
```
Would it make sense to push this check into `getUsableVersion`? I wouldn't consider a negative version "usable."

cmccabe replied:
ok

cmccabe replied:
sounds good

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -502,6 +483,15 @@ private void handleTimedOutRequests(List<ClientResponse> responses, long now) {
     }
 
+    private void handleAbortedSends(List<ClientResponse> responses, long now) {
+        for (Iterator<ClientResponse> iter = abortedSends.iterator();
```
Is this equivalent to this?
```java
responses.addAll(abortedSends);
abortedSends.clear();
```

cmccabe replied:
yeah, let's just do that.

cmccabe replied:
yes, let's replace it with that

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -534,30 +524,29 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
 
-    private void handleApiVersionsResponse(InFlightRequest req, ApiVersionsResponse apiVersionsResponse) {
+    private void handleApiVersionsResponse(List<ClientResponse> responses,
+            InFlightRequest req, long now, ApiVersionsResponse apiVersionsResponse) {
```
nit: indentation

cmccabe replied:
k

cmccabe replied:
fixed

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -534,30 +524,29 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
-                        " required versions: " + "[" + requiredApiVersion.minVersion + ", " + requiredApiVersion.maxVersion + "]" + ".");
-            }
+        if (apiVersionsResponse.errorCode() != Errors.NONE.code()) {
+            log.warn("Node " + node + " got error " + apiVersionsResponse.errorCode() +
```
nit: use string interpolation

cmccabe replied:
yeah

cmccabe replied:
k

ijuma replied:
We have the following code for failed metadata requests:

```java
          if (cluster.isBootstrapConfigured()) {
                int nodeId = Integer.parseInt(destination);
                Node node = cluster.nodeById(nodeId);
                if (node != null)
                    log.warn("Bootstrap broker {}:{} disconnected", node.host(), node.port());
            }
```

Should we be doing similarly for this case (since we now do `ApiVersionsRequest` before `MetadataRequest`)?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -534,30 +524,29 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
+            log.warn("Node " + node + " got error " + apiVersionsResponse.errorCode() +
+                    " when making an ApiVersionsRequest.  Disconnecting.");
+            this.selector.close(node);
+            processDisconnection(responses, node, now);
```
There's a bug in `processDisconnection` (not introduced in this patch) that we ought to fix here. The type of `node` here is a String, but `nodeApiVersions` is keyed by Integer. (The conversion between string and int all over this class always pains me.)

cmccabe replied:
yeah, that is pretty nasty.  I'll fix it.  I wish there were some workaround for this, but it looks like Java Map still always supports put(Object) to allow compatibility with the pre-generics days...

We should start running findBugs to catch things like this

cmccabe replied:
ugh.  That's a nasty one.  I fixed it in the patch.

We should run findBugs...

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -534,30 +524,29 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
+            return;
         }
-        nodeApiVersions.put(Integer.parseInt(node), apiVersionsResponse.apiVersions());
+        int nodexIndex = Integer.parseInt(node);
```
I think `nodeId` is a more accurate name.

cmccabe replied:
k

cmccabe replied:
k

ijuma commented on clients/src/main/java/org/apache/kafka/common/errors/ObsoleteBrokerException.java:
```diff
@@ -0,0 +1,31 @@
+package org.apache.kafka.common.errors;
+
+/**
+ * Indicates that a request cannot be completed because an obsolete broker
```
Nit: is it right to say it's an `obsolete` broker? Seems a bit too strong, obsolete means `no longer produced or used; out of date` whereas it could be a broker that is just 4 months old given our current time-based release schedule.

hachikuji replied:
Maybe `UnsupportedBrokerException`?

cmccabe replied:
obsolete may be a little strong, but it's clear what it means and what the user should do if they want access to the feature.  We could also go with something like "OutdatedBrokerException"?

ijuma replied:
I think I prefer `UnsupportedBrokerVersion`. But it's subjective and `obsolete`/`outdated` are more explicit with regards that you want a newer broker.

Another point: if we think of the protocol as something independent from Kafka (i.e. there could be other implementations), which is something that Jay thinks we should, then `UnsupportedBrokerVersion` also seems better.

cmccabe replied:
From a user point of view, UnsupportedBrokerVersion is that it doesn't make it clear whether the broker version is too new or old.  It also suggests that the whole broker is unsupported, whereas just this one API is unsupported.

> Another point: if we think of the protocol as something independent from Kafka (i.e. there could be other implementations), which is something that Jay thinks we should, then UnsupportedBrokerVersion also seems better.

Hmm.  It seems like we could still have other implementations of the protocol even if we used OutdatedBrokerException / ObsoleteBrokerException?

I'm open to ideas on this.  I would kind of like something that gives the user a clear hint that they need to upgrade...

hachikuji replied:
Another suggestion: `IncompatibleProtocolVersion`?

ijuma replied:
I asked Colin to add this point to the follow-up JIRA.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -42,8 +42,8 @@
 
-    private static final ApiKeys[] ID_TO_TYPE;
-    private static final int MIN_API_KEY = 0;
+    public static final ApiKeys[] ID_TO_TYPE;
```
Do we really need to do this? I'd prefer if we didn't expose a static mutable array outside this class.

cmccabe replied:
ok, we can hide it

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -743,11 +725,13 @@ private long maybeUpdate(long now, Node node) {
                 else
-                    metadataRequest = new MetadataRequest(new ArrayList<>(metadata.topics()));
+                    metadataRequest = new MetadataRequest.Builder(
+                            new ArrayList<>(metadata.topics()));
```
nit: seems like there's enough room on the previous line for this

cmccabe replied:
k

cmccabe replied:
ok

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java:
```diff
@@ -19,29 +19,64 @@
-     * Get an error response for a request for a given api version
+     * Get the version of this AbstractRequest object.
+     */
+    public short getVersion() {
```
Nit: `version`?

cmccabe replied:
k

cmccabe replied:
k

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -584,27 +573,20 @@ private void handleConnections(long now) {
+            this.connectionStates.checkingApiVersions(node);
+            nodesNeedingApiVersionsFetch.add(node);
         }
-        maybeInitiateApiVersionsFetch(now);
```
Seems this was our last usage for the `now` parameter.

cmccabe replied:
k

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -584,27 +573,20 @@ private void handleConnections(long now) {
-                        nextRequestHeader(ApiKeys.API_VERSIONS), ApiVersionsRequest.API_VERSIONS_REQUEST, null);
+                ApiVersionsRequest.Builder apiVersionRequest = new ApiVersionsRequest.Builder();
+                ClientRequest clientRequest = newClientRequest(
+                        node, apiVersionRequest, now, true, null);
```
nit: can this fit on the line above?

cmccabe replied:
k

cmccabe replied:
yeah

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -184,16 +186,21 @@ class RequestSendThread(val controllerId: Int,
+              val version = apiVersion match  {
+                case None => ProtoUtils.latestVersion(apiKey.id)
+                case Some(v) => v
+              }
```
These 3 lines can simply be `val version = apiVersion.getOrElse(ProtoUtils.latestVersion(apiKey,id))`.

cmccabe replied:
nice

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -296,31 +253,76 @@ public void send(ClientRequest request, long now) {
+        try {
+            NodeApiVersions versionInfo = nodeApiVersions.get(Integer.parseInt(nodeId));
+            // Note: if versionInfo is null, we have no server version information.
+            // This would be the case when sending the initial ApiVersionRequest which
```
Seems we can turn this into an assertion when it is not an ApiVersion request.

cmccabe replied:
OK

cmccabe replied:
ok

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -375,14 +383,21 @@ class ControllerBrokerRequestBatch(controller: KafkaController) extends  Logging
       }
 
-      val version = if (controller.config.interBrokerProtocolVersion >= KAFKA_0_10_0_IV1) 2: Short
+      val version :Short  = if (controller.config.interBrokerProtocolVersion >= KAFKA_0_10_0_IV1) 2: Short
```
Nit: it should be `val version: Short` (space after the colon, not before).

cmccabe replied:
k

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -502,6 +483,15 @@ private void handleTimedOutRequests(List<ClientResponse> responses, long now) {
             metadataUpdater.requestUpdate();
     }
 
+    private void handleAbortedSends(List<ClientResponse> responses, long now) {
```
`now` is never used.

cmccabe replied:
removed

cmccabe replied:
removed

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -391,7 +406,9 @@ class ControllerBrokerRequestBatch(controller: KafkaController) extends  Logging
-          new UpdateMetadataRequest(version, controllerId, controllerEpoch, partitionStates.asJava, liveBrokers.asJava)
+          new UpdateMetadataRequest.Builder(
+              controllerId, controllerEpoch, partitionStates.asJava, liveBrokers.asJava).
+              setVersion(version)
```
Looks like the code for both `Builder(...)` is the same, so maybe we can just do the `liveBrokers` bit in the `if/else`.

cmccabe replied:
k

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -375,14 +383,21 @@ class ControllerBrokerRequestBatch(controller: KafkaController) extends  Logging
+          // Version 0 of UpdateMetadataRequest only supports PLAINTEXT.
+          val liveBrokers = controllerContext.liveOrShuttingDownBrokers.map { broker =>
+            val node = broker.getNode(SecurityProtocol.PLAINTEXT)
+            val endPoints = Map{ SecurityProtocol.PLAINTEXT -> new EndPoint(node.host(), node.port()) }
```
Nit: the map creation is usually done like `Map(SecurityProtocol.PLAINTEXT -> new EndPoint(node.host, node.port))`.

cmccabe replied:
k

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -95,23 +91,11 @@
-        this(NO_METADATA_UPDATER, metadata, selector, clientId, maxInFlightRequestsPerConnection,
-                reconnectBackoffMs, socketSendBuffer, socketReceiveBuffer, requestTimeoutMs, time, requiredApiVersions);
-    }
+    private final List<ClientResponse> abortedSends = new LinkedList<ClientResponse>();
```
nit: you can leave out the type on the right-hand side.

cmccabe replied:
k

cmccabe replied:
k

ijuma replied:
Any reason we are using `LinkedList` instead of `ArrayList`? We seem to be adding to the end, iterating and then clearing (after Jason's suggestion), so `ArrayList` seems like the better option.

ijuma commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -288,11 +292,10 @@ class ReplicaFetcherThread(name: String,
-    new FetchRequest(request)
+    val requestBuilder = new JFetchRequest.Builder(maxWait, minBytes, requestMap).
+        setReplicaId(replicaId).setMaxBytes(maxBytes)
+    requestBuilder.setVersion(fetchRequestVersion)
```
Nit: is there a reason why you didn't chain this one too?

cmccabe replied:
setVersion is a method on the base class, AbstractRequest#Builder.  So it returns an instance of AbstractRequest#Builder rather than an instance of the derived class FetchRequest#Builder, which is awkward for Scala's type inference.

ijuma replied:
Oh, I see. The compiler is correct in not allowing that then. One way to fix that is to change the definition of `Builder` to be:

```java
public static abstract class Builder<T extends AbstractRequest, B extends Builder<T, B>> {
  ...
  public B setVersion(short version) {
  ...
```

But this makes it more awkward to use the abstract builder directly due to the additional type parameter (in Scala we could use type members instead). An alternative is to override `setBuilder` in each builder and use the more specific type there. It's a bit annoying, but builders are supposedly used more times than they are defined.

ijuma commented on core/src/main/scala/kafka/utils/NetworkClientBlockingOps.scala:
```diff
@@ -105,10 +105,9 @@ class NetworkClientBlockingOps(val client: NetworkClient) extends AnyVal {
     pollContinuously { responses =>
       val response = responses.find { response =>
-        response.requestHeader.correlationId == request.header.correlationId
+        response.requestHeader().correlationId() == request.correlationId
```
Seems like the `()` introduced here are not needed.

cmccabe replied:
k

hachikuji reviewed (commented):
Some comments in `NetworkClient`. Logical flow makes sense overall.

ijuma reviewed (commented):
A few more minor comments. Still going through it, but sending these to reduce latency.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:
```diff
@@ -86,25 +85,19 @@ public ConsumerNetworkClient(KafkaClient client,
-    }
-
-    private RequestFuture<ClientResponse> send(Node node,
                                                ApiKeys api,
```
`api` no longer used. Indentation needs fixing also.

hachikuji replied:
There are a few javadoc references that need fixing as well.

cmccabe replied:
k

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -547,16 +583,29 @@ public void onFailure(RuntimeException e) {
                 .compose(new RequestFutureAdapter<ClientResponse, Map<TopicPartition, OffsetAndTimestamp>>() {
                     @Override
                     public void onSuccess(ClientResponse response, RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> future) {
+                        if (response.versionMismatch() != null) {
```
I think it might be better to move this check into `ConsumerNetworkClient.RequestFutureCompletionHandler` to ensure that we don't forget any checks. Also `onFailure` seems like a more appropriate callback for that case,

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -579,11 +629,34 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
+                    } else {
+                        offset = partitionData.offsets.get(0);
+                    }
+                    log.debug("handleListResponse: handling v0 response for " + topicPartition +
```
nit: string interpolation is preferred. another one below. Also, not sure about the convention of including the function name. Wouldn't the logger do that for us?

cmccabe replied:
Yeah, let's leave the function name out.  Will use string interpolation

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -579,11 +629,34 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
-                    offsetAndTimestamp = new OffsetAndTimestamp(partitionData.offset, partitionData.timestamp);
-                log.debug("Fetched {} for partition {}", offsetAndTimestamp, topicPartition);
-                timestampOffsetMap.put(topicPartition, offsetAndTimestamp);
+                if (partitionData.offsets != null) {
```
In this case, we're using the presence of the `offsets` field to indirectly determine that the version that was used was 0. I'm not sure we'll always have something so convenient, so I've been wondering if we need a way to determine the version that was used more directly. For example, we could have `ClientResponse` or even `AbstractResponse` include a field for the version.

cmccabe replied:
I think AbstractResponse should have a version number in it, since the data it contains varies based on the version.  Sometimes the interpretation of the data varies based on version as well.  It's a property of the response and it belongs in the response object.  But hopefully we can hold off on that for now since this patch is kinda big as-is...

hachikuji replied:
I doubt it would add much to the size of this patch if you want to do it here. A follow-up would work as well. I'm anxious to have a good general approach in place so that we have clear patterns to follow going forward.

cmccabe replied:
Yeah, it would be good to have a clear pattern here.  Let's do it in a follow-on.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -518,12 +552,14 @@ else if (strategy == OffsetResetStrategy.LATEST)
                         public void onSuccess(Map<TopicPartition, OffsetAndTimestamp> value) {
                             synchronized (listOffsetRequestsFuture) {
-                                fetchedTimestampOffsets.putAll(value);
+                                for (Map.Entry<TopicPartition, OffsetAndTimestamp> entry: value.entrySet()) {
```
Why was this changed?

cmccabe replied:
Good catch.  I made some other edits to the for loop, but they're no longer necessary now.  Reverted

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -331,31 +332,59 @@ else if (error.exception() instanceof RetriableException)
+        return getOffsetsByTimesImpl(timestampsToSearch, timeout, true);
+    }
+
+    private Map<TopicPartition, OffsetAndTimestamp> getOffsetsByTimesImpl(
```
The `Impl` suffix is a little unconventional. Can we just use `getOffsetsByTimes`?

cmccabe replied:
I wanted to avoid confusion with the public API of that name.  How about "retrieveOffsetsByTimes"?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -579,11 +629,34 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
+                    // Handle v0 response
+                    long offset;
+                    if (partitionData.offsets.size() > 1) {
+                        throw new IllegalStateException("Unexpected partitionData response of length " +
```
I think we need to pass this exception to the future instead of raising. Would be good to have a test case if we don't already.

hachikuji replied:
@cmccabe Can you address this comment please?

cmccabe replied:
done

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -331,31 +332,59 @@ else if (error.exception() instanceof RetriableException)
+            client.poll(future);
+            if (future.succeeded()) {
+                Map<TopicPartition, OffsetAndTimestamp> ret = future.value();
+                return ret.get(partition).offset();
```
We didn't have it before, but maybe we should add a null check here for more resilience in the future.

cmccabe replied:
Hmm, I didn't see any code path where the return value could be null, or the value of the future itself, or the keys or values of the map.  Maybe an NPE is ok in that case?

ijuma replied:
I think @hachikuji is thinking of the case where `ret.get(partition)` returns `null`. Not sure if we are enforcing that elsewhere though.

cmccabe replied:
Ah, good point.  Let me add a check for that just to be safe.  After all, this data is coming over the wire, so we should not unconditionally trust it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -649,10 +722,12 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
             Node node = entry.getKey();
-            FetchRequest fetch = new FetchRequest(this.maxWaitMs, this.minBytes, this.maxBytes, entry.getValue());
+            FetchRequest.Builder fetch = new FetchRequest.Builder(
+                    this.maxWaitMs, this.minBytes, entry.getValue()).
```
I'm a little unclear on the pattern for which fields are included in the builder constructor and which are included through methods. I thought perhaps it would be the required arguments included in the constructor, but we didn't pass the timestamps to query in the `ListOffsetRequest` above, which seems required.

cmccabe replied:
You are right that the general pattern is that constructor arguments are required, other things are optional.  Ideally we would have ListOffsetRequest#Builder(Map<TopicPartition, PartitionData>) and ListOffsetRequest#Builder(Map<TopicPartition, Long>), but we can't do that because of a Java limitation... both constructors would have the same type once type erasure has been performed, which is not allowed.

ijuma replied:
I submitted a PR that removes unused setters, makes fields final and tries to make things a bit more regular. Makes it a bit simpler, but more could be done probably.

hachikuji reviewed (commented):
Some comments in `Fetcher`.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/KafkaClient.java:
```diff
@@ -112,24 +111,21 @@
-     */
-    public void wakeup();
-
+    public ClientRequest newClientRequest(String nodeId, AbstractRequest.Builder requestBuilder,
```
I think newRequest would raise the question of whether we were creating a new AbstractRequest or a new ClientRequest... or maybe that's just me.  Thoughts?

cmccabe commented on clients/src/main/java/org/apache/kafka/common/errors/ObsoleteBrokerException.java:
```diff
@@ -0,0 +1,31 @@
+package org.apache.kafka.common.errors;
+
+/**
+ * Indicates that a request cannot be completed because an obsolete broker
```
I agree that "obsolete" might be a bit of an exaggeration, but it gets the point across very nicely.  Maybe we could go with "OutdatedBrokerException" if that sounds better?  I'm open to suggestions

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/GroupCoordinatorRequest.java:
```diff
@@ -16,32 +16,58 @@
+        @Override
+        public String toString() {
+            StringBuilder bld = new StringBuilder();
+            bld.append("{type: GroupCoordinatorRequest, groupId: ");
```
nit: Would be nice to be consistent on the use of braces or parenthesis. I think we are trying to encourage the latter.

cmccabe replied:
Braces seem a little more familiar to me because of JSON.  But if parentheses are the way to go, I can change it.  It's a big change though since all the builders have a toString.  Are we sure we want the parens?

ijuma replied:
There are indeed a few ways to do this. We tend to follow the `Foo(a="a",b="b")` model and this is what public classes like `ProducerRecord` and `ConsumerRecord` do.

Also, it's worth saying that, outside of loops, `StringBuilder` is no better than String concatenation with regards to performance. Up to java 8, javac will do the StringBuilder thing itself. From Java 9, invokedynamic will be used to allow the runtime to choose the best strategy (http://openjdk.java.net/jeps/280).

My suggestion is to clean this up in a subsequent PR. It can even be done after the feature freeze.

junrao commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -296,31 +254,78 @@ public void send(ClientRequest request, long now) {
+                throw new IllegalStateException("Attempt to send a request to node " + nodeId + " which is not ready.");
+        }
+        AbstractRequest request = null;
+        AbstractRequest.Builder builder = clientRequest.requestBuilder();
```
When the transport is SSL, the client can't send requests immediately after the socket is connected. We have to wait for the SSL handshake to complete. So, perhaps the selector needs to further return the transport.ready() state in KafkaChannel to NetworkClient.

ijuma replied:
@cmccabe, have we tested with SSL enabled?

cmccabe replied:
Hmm.  canSendRequest should be making sure that the SSL handshake is complete, right?

Call stack:
```
NetworkClient#doSend
NetworkClient#canSendRequest
Selector#isChannelReady
KafkaChannel#ready
```

KafkaChannel#ready has:
```
    public boolean ready() {
        return transportLayer.ready() && authenticator.complete();
    }
```

cmccabe replied:
@ijuma: I'll parameterize our ducktape tests for client compat with SSL

junrao replied:
In the patch, it seems that once a channel is connected, but is not necessarily ready, NetworkClient can start issuing ApiRequest. This can fail if transport is SSL and transportLayer.ready() is false. It's also too late to only start issuing ApiRequest after channel is ready since the ApiRequest should be sent before SASL handshake (SASL handshake completes when authenticator.complete() is true). 

What the NetworkClient is supposed to do is only start issuing ApiRequest after transportLayer.ready() is true. It's just that currently, this state is no propagated by Selector.

ijuma replied:
@junrao I'm not sure what you mean when you say that the patch starts issuing ApiVersions request before transport layer is ready:

```java
private void handleInitiateApiVersionRequests(long now) {
        Iterator<String> iter = nodesNeedingApiVersionsFetch.iterator();
        while (iter.hasNext()) {
            String node = iter.next();
            if (selector.isChannelReady(node) && inFlightRequests.canSendMore(node)) {
                log.debug("Initiating API versions fetch from node {}.", node);
                ApiVersionsRequest.Builder apiVersionRequest = new ApiVersionsRequest.Builder();
                ClientRequest clientRequest = newClientRequest(node, apiVersionRequest, now, true, null);
                doSend(clientRequest, true, now);
                iter.remove();
            }
        }
    }
```
Isn't that OK (we call `selector.isChannelReady(node)`)?

The point about SASL. The outcome of the discussion in the PR for KAFKA-3600 was that we don't have to send the api versions request before the SASL request for now since there's only one version of the SaslHandshake request. If we ever introduce an additional version, we can change the client at the same time. It seems like there's no downside to this since brokers will have to keep supporting the old sasl handshake request anyway. Do you feel differently?

junrao replied:
@ijuma : Yes, I missed the selector.isChannelReady(node) check. So, this is fine. About the SASL, that's reasonable. We can keep the changes in the PR as they are.

junrao commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java:
```diff
@@ -156,8 +156,9 @@ public void authenticate() throws IOException {
-                currentRequestHeader = new RequestHeader(ApiKeys.SASL_HANDSHAKE.id, clientId, correlationId++);
                 SaslHandshakeRequest handshakeRequest = new SaslHandshakeRequest(mechanism);
+                currentRequestHeader = new RequestHeader(ApiKeys.SASL_HANDSHAKE.id,
+                        handshakeRequest.getVersion(), clientId, correlationId++);
```
Using the current version is ok for now since there is currently only one version of SaslHandshakeRequest. It would be useful to add a comment that when there are new versions of SaslHandshakeRequest, we need to select the version based on the result of ApiRequest accordingly.

ijuma replied:
@junrao, there is a comment already a couple of lines above (I asked Ashish to add it as part of KAFKA-3600):

```
// When multiple versions of SASL_HANDSHAKE_REQUEST are to be supported,
// API_VERSIONS_REQUEST must be sent prior to sending SASL_HANDSHAKE_REQUEST to
// fetch supported versions.
```

Maybe we can move the comment so that it's closer to the request creation.

junrao reviewed (commented):
Thanks for the patch. Left a couple of comments. 

Also, would it be useful to log a warning when the client needs to issue an old version of the request since certain feature could be lost (e.g. timestamp)?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -349,19 +347,20 @@ private void sendProduceRequest(long now, int destination, short acks, int timeo
 
         client.send(clientRequest, now);
-        log.trace("Sent produce request to {}: {}", nodeId, produceRequest);
+        log.trace("Sent produce request to {}: {}", nodeId, requestBuilder);
```
We seem to be missing the version mismatch check in `handleProduceResponse`.

ijuma replied:
@cmccabe, this one seems important to resolve before merging.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -296,31 +253,76 @@ public void send(ClientRequest request, long now) {
+            // that we can send data.  If the request is internal, we trust
+            // that that internal code has done this validation.  Validation
+            // will be slightly different for some internal requests (for
+            // example, ApiVersionsRequests can be sent prior to being in
```
Hmm.  Both the MetadataUpdateRequest and the ApiVersionsRequest do their own checks before invoking doSend, so it would be duplicated work.

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -584,27 +573,20 @@ private void handleConnections(long now) {
+            this.connectionStates.checkingApiVersions(node);
+            nodesNeedingApiVersionsFetch.add(node);
         }
-        maybeInitiateApiVersionsFetch(now);
```
k

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -456,6 +456,20 @@ public static String formatAddress(String host, Integer port) {
+    public static <K, V> String join(Map<K, V> map,
+                                     String keyValueSeparator, String elementSeperator) {
+        StringBuilder bld = new StringBuilder();
+        bld.append("{");
```
It's a bit odd that we add a hardcoded prefix and suffix in a generic `join` method. Typically, these would be parameters and the default case would be no prefix or suffix (consistent with `join` that takes a `collection`).

cmccabe replied:
renamed to Utils#mkString and made this optional with function overloads.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/Utils.java:
```diff
@@ -456,6 +456,20 @@ public static String formatAddress(String host, Integer port) {
+        StringBuilder bld = new StringBuilder();
+        bld.append("{");
+        String prefix = "";
+        for (Map.Entry<K, V> entry: map.entrySet()) {
```
Nit: space missing before `:`.

cmccabe replied:
k

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -36,22 +36,22 @@ class MetadataRequestTest extends BaseRequestTest {
   @Test
   def testClusterIdWithRequestVersion1() {
-    val v1MetadataResponse = sendMetadataRequest(MetadataRequest.allTopics, 1)
+    val v1MetadataResponse = sendMetadataRequest(MetadataRequest.allTopics(1.toShort))
```
Nit: `1: Short` is a little better because it will fail to compile if the conversion is not possible where `toShort` will happily overflow. Doesn't make a difference in these particular cases though.

cmccabe replied:
ok.  will leave it as toShort here since the typecast is kind of annoying visually
(edit: it's an ascription) ;)

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -177,79 +177,87 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
   private def createListOffsetsRequest = {
-    new requests.ListOffsetRequest(Map(tp -> new ListOffsetRequest.PartitionData(0, 100)).asJava)
+    new requests.ListOffsetRequest.Builder().setTargetTimes(
+      Map(tp -> java.lang.Long.valueOf(0)).asJava).
```
Nit: `0: java.lang.Long` is a tiny bit shorter.

cmccabe replied:
k

ijuma commented on clients/src/main/java/org/apache/kafka/clients/KafkaClient.java:
```diff
@@ -112,24 +111,21 @@
-
+    public ClientRequest newClientRequest(String nodeId, AbstractRequest.Builder requestBuilder,
+                                   long createdTimeMs, boolean expectResponse,
+                                   RequestCompletionHandler callback);
```
I think it would be worth adding an overload without the `callback` as it's optional and many callers don't need it.

ijuma replied:
I've done this in my PR.

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -184,16 +186,21 @@ class RequestSendThread(val controllerId: Int,
+              }
+              val clientRequest = networkClient.newClientRequest(
+                  brokerNode.idString, requestBuilder, time.milliseconds(), true, null)
+              val abstractRequest = clientRequest.requestBuilder().build()
```
This is unused.

ijuma replied:
Removed in my PR.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/ListOffsetRequest.java:
```diff
@@ -137,8 +210,8 @@ private ListOffsetRequest(int replicaId, Map<TopicPartition, ?> targetTimes, int
-        super(struct);
+    public ListOffsetRequest(Struct struct, short versionId) {
+        super(struct, versionId);
         Set<TopicPartition> duplicatePatitions = new HashSet<>();
```
Typo, `r` missing.

ijuma replied:
Fixed in my PR.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java:
```diff
@@ -118,11 +101,14 @@ public void testClose() {
-        ClientRequest request = new ClientRequest(node.idString(), time.milliseconds(), true, reqHeader, produceRequest, null);
+        ProduceRequest.Builder builder =
+            new ProduceRequest.Builder((short) 1, 1000,
+                Collections.<TopicPartition, MemoryRecords>emptyMap());
```
Nit: I think this is harder to read by being split into 3 lines like this. Looks like a staircase. ;) There's one other case like that.

cmccabe replied:
k

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/ControlledShutdownRequest.java:
```diff
@@ -60,10 +80,12 @@ public int brokerId() {
     }
 
     public static ControlledShutdownRequest parse(ByteBuffer buffer, int versionId) {
```
Change `versionId` to `short` like you did in other places?

cmccabe replied:
I wanted to do this, but it got messy.  The problem is that there are a bunch of unit tests that are using reflection to call these methods-- and also some scala code.  I think we should have a follow-on JIRA for converting the parse() methods to use 'short' for version.

ijuma reviewed (commented):
Left a few more minor comments. About the builders, there are a couple of things that I am unsure about:

1. In some classes, we duplicate a bunch of fields between the builder and the request class.
2. We pass around these mutable builders across layers, which is not as nice as our previously immutable request classes. Do we ever want to change anything besides version?

One alternative would be something along the lines of:

```java
public static abstract class Builder<T extends AbstractRequest> {

        private final ApiKeys apiKey;
        private final short version;

        public Builder(ApiKeys apiKey, short version) {
            this.apiKey = apiKey;
            this.version = version;
        }

        public Builder(ApiKeys apiKey) {
            this(apiKey, ProtoUtils.latestVersion(apiKey.id));
        }

        public abstract T build();
        public abstract T build(short version);
    }
```

That would make the builders immutable and eliminate issue 2.

For 1, we could introduce `Params` classes where they made sense. Because they're immutable, we don't necessarily need to do field by field copy. We'd have to check if this makes things better or worse though.

I don't think we need to do the above in this PR, but raising it as an idea to collect feedback.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java:
```diff
@@ -83,16 +83,89 @@ public PartitionData(long offset, long timestamp, String metadata) {
+                    return new OffsetCommitRequest(groupId, generationId, memberId,
+                            offsetData);
+                case 2:
+                    return new OffsetCommitRequest(groupId, generationId, memberId,
```
It looks like we only need to support `2` and above: https://github.com/apache/kafka/blob/0.10.0/clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java

cmccabe replied:
The other versions are used in unit tests.  The server also can handle them to support older clients.

ijuma replied:
The server doesn't use the builder so I think that would be OK. But let's leave it for now.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java:
```diff
@@ -42,12 +37,49 @@
+
+        @Override
+        public OffsetFetchRequest build() {
+            return new OffsetFetchRequest(groupId, partitions, version());
```
Should we be validating that version is always `1`? Version `0` would probably mean that something is wrong.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java:
```diff
@@ -83,16 +83,89 @@ public PartitionData(long offset, long timestamp, String metadata) {
+                case 2:
+                    return new OffsetCommitRequest(groupId, generationId, memberId,
+                            retentionTime, offsetData, version);
+                default:
```
In other request classes we don't throw an exception in case the version is higher than expected. We should be consistent. I think this approach is good because it forces people to consider how to handle a new version when it's introduced.

cmccabe replied:
Hmm.  Just to clarify, you think that throwing the exception when we see an unexpected version is the right way to go?  I can add that to the other requests if so.

ijuma replied:
Yes, but let's consider that in a separate PR.

cmccabe replied:
sounds good

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/JoinGroupRequest.java:
```diff
@@ -62,51 +60,103 @@ public ByteBuffer metadata() {
+        public JoinGroupRequest build() {
+            short version = version();
+            if (version < 1) {
+                rebalanceTimeout = sessionTimeout;
```
I think we should set it to `-1` in this case to make it clear that it won't be used as we have the following:

```java
if (version >= 1) {
    struct.set(REBALANCE_TIMEOUT_KEY_NAME, rebalanceTimeout);
}
```

cmccabe replied:
k

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java:
```diff
@@ -74,14 +127,15 @@ public AbstractResponse getErrorResponse(int versionId, Throwable e) {
             default:
                 throw new IllegalArgumentException(String.format("Version %d is not valid. Valid versions for %s are 0 to %d",
-                        versionId, this.getClass().getSimpleName(), ProtoUtils.latestVersion(ApiKeys.METADATA.id)));
+                        getVersion(), this.getClass().getSimpleName(), ProtoUtils.latestVersion(ApiKeys.METADATA.id)));
```
We have 3 `getVersion` calls. One should be enough.

ijuma replied:
Changed in my PR.

ijuma commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -258,15 +258,19 @@ class ReplicaFetcherThread(name: String,
+    val requestBuilder = if (brokerConfig.interBrokerProtocolVersion >= KAFKA_0_10_1_IV2) {
         val partitions = Map(topicPartition -> (earliestOrLatest: java.lang.Long))
-        (new ListOffsetRequest(partitions.asJava, consumerId), 1)
+        (new ListOffsetRequest.Builder().
```
Parenthesis are not needed here or the other branches since a single value is being returned now.

cmccabe replied:
k

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -88,43 +85,64 @@ public TopicAndPartitionData(String topic) {
+            return this;
+        }
+
+        public LinkedHashMap<TopicPartition, PartitionData> getFetchData() {
```
Nit: should be `fetchData()`.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -0,0 +1,116 @@
+        this.apiVersions = apiVersions;
+        this.usableVersions = new short[ApiKeys.ID_TO_TYPE.length];
+        Arrays.fill(usableVersions, (short) -1);
+        for (ApiVersionsResponse.ApiVersion apiVersion: apiVersions) {
```
Nit: there should be a space before `:`.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    public short getUsableVersion(short apiKey) {
```
Personally, I think `Short` and `null` for no usable version is less error-prone (you'll get a NPE if you accidentally try to use it instead of ending with a `-1` somewhere it should not be).

cmccabe replied:
I fixed it to be an exception instead

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -0,0 +1,116 @@
+        // unknown, which may happen when either the client or server is newer.
+        //
+        // This toString method is relatively expensive, so you probably want to avoid
+        // calling it unless debug logging is turned on.
```
Seems like this comment should be in the method documentation.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -0,0 +1,116 @@
+    private final Collection<ApiVersionsResponse.ApiVersion> apiVersions;
+
+    // An array of the usable versions of each API, indexed by ApiKeys ID.
+    private final short[] usableVersions;
```
Why not use `EnumMap<ApiKeys, Short>`? It would be more readable IMO.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -0,0 +1,116 @@
+        Arrays.fill(usableVersions, (short) -1);
+        for (ApiVersionsResponse.ApiVersion apiVersion: apiVersions) {
+            int index = apiVersion.apiKey;
+            if ((index < 0) || (index >= usableVersions.length)) {
```
Can this ever be less than 0?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -0,0 +1,116 @@
+                usableVersions[index] = Protocol.CURR_VERSION[index];
+            } else {
+                usableVersions[index] = apiVersion.maxVersion;
+            }
```
`Math.min` and 4 lines become 1?

cmccabe replied:
For some reason, there is no Math.min(short, short), only Math.min(int, int).  I could use the int version with a typecast, but I thought the cast was ugly.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -0,0 +1,116 @@
+            prefix = ", ";
+        }
+        bld.append("}");
+        return bld.toString();
```
This `toString` is quite long. :) It would be nice to break it down into smaller methods so that it's easier to understand.

cmccabe replied:
k

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -296,31 +254,78 @@ public void send(ClientRequest request, long now) {
+                short version = versionInfo.getUsableVersion(clientRequest.apiKey().id);
+                if (log.isTraceEnabled()) {
+                    log.trace("When sending message of type {} to node {}, the best usable " +
+                            "version is {}", clientRequest.apiKey(), nodeId, version);
```
It would be good to include the most current version in this message as well. Also, as Jun said it may be a good to log at a higher level if we are downgrading a request. Not sure what Jun had in mind, but maybe `debug`?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -584,27 +576,20 @@ private void handleConnections(long now) {
             log.debug("Completed connection to node {}", node);
```
`now` is no longer used

cmccabe replied:
k

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -584,27 +576,20 @@ private void handleConnections(long now) {
     }
 
-    private void maybeInitiateApiVersionsFetch(long now) {
-        if (this.requiredApiVersions == null)
```
We are now doing `ApiVersions` unconditionally. This means that 0.10.2 brokers won't be able to talk to 0.9.0.x brokers, right? We shouldn't do that.

cmccabe replied:
That's a good point.  What do you suggest?  One approach is to (re)add a constructor parameter to NetworkClient that will allow the brokers to bypass sending ApiVersions.

hachikuji replied:
Sounds reasonable to me.

cmccabe replied:
OK, I added a constructor parameter and a unit test.  Thanks for catching this!

ijuma reviewed (commented):
A few more comments. The important one is regarding ApiVersions being required for brokers.

cmccabe committed:
OffsetAndTimestamp: improve comment

cmccabe committed:
ClientRequest: re-order the elements printed in toString

cmccabe committed:
Some whitespace cleanups

cmccabe committed:
change AbstractRequest#getVersion to AbstractRequest#version

cmccabe committed:
NodeApiVersions: some improvements

* don't require ApiKeys#ID_TO_APIKEY to be public
* getUsableVersion should throw exception when there is no usable version

cmccabe committed:
ControllerChannelManager: some style and whitespace fixes

cmccabe committed:
NetworkClientBlockingOps: leave out unnecessary parentheses

cmccabe committed:
NetworkClient: some fixes

* Remove unnecessary parentheses
* handleAbortedSends doesn't need to take a 'now' parameter
* fix a place where we were removing from nodeApiVersions using the wrong type (String instead of Integer)
* convert log string concatenation to using placeholders
* rename 'int nodeIndex' -> 'int nodeId'

cmccabe committed:
Some refactors

* remove a superfluous client.send() parameter
* getOffsetsByTimesImpl -> retrieveOffsetsByTimes

cmccabe committed:
NetworkClient: remove unused 'now' parameter from handleConnections

cmccabe commented:
I posted an update with a lot of fixes.  I still need to address:
* Brokers should not send ApiVersionsRequest
* Implement pre-KIP-74 message-too-big behavior
* More centralized handling of versionMismatch exception
* some other stuff

cmccabe committed:
Fetcher.java: add a check for null in Fetcher#listOffset

hachikuji commented:
@ijuma I think your builder suggestion seems worth investigating.. If we continue with the model of delayed version determination, I was considering whether we could decouple the `Struct` representation from the request objects. Haven't really thought this through, but I was thinking something like this:

```java
public abstract class AbstractRequest {
  abstract Struct toStruct(short version);
}
``` 

That way the request objects remain immutable. I think you also had a suggestion which allowed us to construct the right version up-front.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -95,23 +92,11 @@
-        this(NO_METADATA_UPDATER, metadata, selector, clientId, maxInFlightRequestsPerConnection,
-                reconnectBackoffMs, socketSendBuffer, socketReceiveBuffer, requestTimeoutMs, time, requiredApiVersions);
-    }
+    private final List<ClientResponse> abortedSends = new LinkedList<>();
```
Note there was a previous comment about using `ArrayList` instead of `LinkedList`.

cmccabe replied:
Oops, sorry, I meant to reply but I closed the browser window.  ArrayList seems like overkill here since we don't expect a lot of aborted sends to happen.  Also, the memory consumed by ArrayList only grows and never shrinks (unless you call trimToSize), which seems like a bad choice here.

ijuma replied:
Well, the `ArrayList` array would expand to 10 elements the first time an aborted send is added. And then, it will probably never grow beyond that. LinkedList in Java is just not a very good collection in general. Each node has 3 references and there is a node instance per item. A funny tweet by Joshua Bloch:

"@jerrykuch @shipilev @AmbientLion Does anyone actually use LinkedList?  I wrote it, and I never use it."
https://twitter.com/joshbloch/status/583813919019573248

cmccabe committed:
Utils.java: whitespace fix

cmccabe committed:
AbstractRequest#Builder: make toString methods more consistent and use parentheses not braces

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -382,33 +379,30 @@ class ControllerBrokerRequestBatch(controller: KafkaController) extends  Logging
         }
+        new UpdateMetadataRequest.Builder(
+            controllerId, controllerEpoch, partitionStates.asJava, liveBrokers.asJava).
+            setVersion(0)
```
This should be `setVersion(version)` I think.

cmccabe committed:
Rename Utils#join(Map...) to Utils#mkString

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -296,31 +254,69 @@ public void send(ClientRequest request, long now) {
+        } catch (ObsoleteBrokerException | UnsupportedVersionException e) {
+            // If the version is not supported, skip sending the request over the wire.
+            // Instead, simply add it to the local queue of aborted requests.
+            log.debug("Version mismatch when attempting to send " + clientRequest.toString() + " to " +
```
nit: string interpolation

cmccabe replied:
k

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -534,30 +514,29 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
-            }
+        if (apiVersionsResponse.errorCode() != Errors.NONE.code()) {
+            log.warn("Node {} got error {} when making an ApiVersionsRequest.  Disconnecting.",
+                    node, apiVersionsResponse.errorCode());
```
Probably more helpful to convert the error code to an instance of `Errors` and use `message()`. Even better would be to have `ApiVersionsResponse` use a field of type `Errors` instead of a short.

cmccabe replied:
Yeah, let's use Errors to get the error enum.  Can refactor further later

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -534,30 +514,29 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
         this.connectionStates.ready(node);
-        log.debug("Recorded API versions for node {}: {}", node, apiVersionsResponse.apiVersions());
+        if (log.isDebugEnabled()) {
+            log.debug("Recorded API versions for node {}: {}", node, nodeVersionInfo.toString());
```
nit: the `toString()` is not needed, right?

cmccabe replied:
true, it should be called implicitly

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -470,7 +445,7 @@ public static AbstractResponse parseResponse(ByteBuffer responseBuffer, RequestH
     private void processDisconnection(List<ClientResponse> responses, String nodeId, long now) {
         connectionStates.disconnected(nodeId, now);
-        nodeApiVersions.remove(nodeId);
+        nodeApiVersions.remove(Integer.valueOf(nodeId));
```
I was thinking about this a little more. I wonder if we could just change the key type to string to make it consistent with `nodesNeedingApiVersionsFetch`?

cmccabe replied:
ok

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java:
```diff
@@ -0,0 +1,127 @@
+        }
+    }
+
+    public short getUsableVersion(short apiKey) {
```
Could we have this accept an `ApiKeys` instance instead of short? Then we wouldn't need the check for a negative api key. Also would be nice to document what exactly this method is returning. Is it the largest version supported by both the client and server?

cmccabe replied:
good ideas.  done

cmccabe committed:
ClientUtils, NodeApiVersions: fix checkstyle

cmccabe committed:
NetworkClient: brokers should not send ApiVersionRequest; only clients should

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -579,11 +626,33 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
+                        timestampOffsetMap.put(topicPartition, offsetAndTimestamp);
+                    }
+                } else {
+                    // Handle v1 and later reponse
```
typo: "reponse"

cmccabe replied:
k

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -547,16 +580,29 @@ public void onFailure(RuntimeException e) {
                 .compose(new RequestFutureAdapter<ClientResponse, Map<TopicPartition, OffsetAndTimestamp>>() {
                     @Override
                     public void onSuccess(ClientResponse response, RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> future) {
+                        if (response.versionMismatch() != null) {
```
I think I had a comment before about moving this check into `ConsumerNetworkClient.RequestFutureCompletionHandler`.

cmccabe replied:
Right, I have been looking at that.  Thinking of doing it in a follow-on JIRA since I will need to take a deeper look at the RPC paths when doing that

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -547,16 +580,29 @@ public void onFailure(RuntimeException e) {
+
+        // If we need a timestamp in the response, the minimum RPC version we can send is v1.
+        // Otherwise, v0 is OK.
+        builder.setMinVersion(requireTimestamp ? (short) 1 : (short) 0);
```
To be clear, if the broker doesn't support this then we'll raise `ObsoleteBrokerException`? I wonder if that handling is consistent with the behavior when the broker is up to date, but has an old message format version. It seems our handling of the `UNSUPPORTED_FOR_MESSAGE_FORMAT` is to add a null entry to the result. Maybe we should do the same?

cmccabe replied:
I don't understand what you mean by "the broker is up to date, but has an old message format version".

hachikuji replied:
I'm thinking of the case where the broker doesn't support v1 of ListOffsets. For this case, I think we currently raise `ObsoleteBrokerException`. I am questioning whether it would be more consistent to return a null entry in this case in the result of `offsetsForTimes`. Currently it is possible for the broker to support the new api version, but not the message format version which is needed to answer the query. In this case, we return a null entry.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndTimestamp.java:
```diff
@@ -21,7 +21,10 @@
- * Both offset and timestamp are non-negative.
+ * Offset must be non-negative.
+ *
+ * The timestamp should never be negative, unless it is invalid.  This could happen when
```
Would we ever return a negative offset to the user?

cmccabe replied:
no

ijuma replied:
We should not have it in the javadoc then as this is a class that is exposed to users.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -380,16 +379,16 @@ public void onFailure(RuntimeException e) {
-        log.debug("Sending JoinGroup ({}) to coordinator {}", request, this.coordinator);
-        return client.send(coordinator, ApiKeys.JOIN_GROUP, request)
+        log.debug("Sending JoinGroup ({}) to coordinator {}",
+                requestBuilder, this.coordinator);
```
nit: move to previous line

cmccabe replied:
k

hachikuji reviewed (commented):
Next round of comments.

cmccabe commented:
* Brokers no longer send ApiVersionsRequest, to allow inter-broker communication with versions before 0.10
* Addressed some style comments

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -807,13 +807,15 @@ public InFlightRequest(RequestHeader header,
-
     }
 
+    public boolean discoverPeerVersions() {
```
Seems the only usage is internal to `NetworkClient`. Maybe we could just use the variable directly?

ijuma replied:
Also, is `discoverPeerVersions` the right name? Seems like it should be `discoverBrokerVersions` as `NetworkClient` is also used by clients.

cmccabe commented:
@hachikuji, @ijuma : unfortunately github doesn't let me post comments on the thread you have going about refactoring AbstractRequest.  So I'll comment here.

There are really two questions here:
1. Should the AbstractRequest objects represent serialized data or unserialized data?
2. Can we know the version of the request that we need before we send it to NetworkClient

For #1, currently, AbstractRequest objects represent serialized data, so they inherently have a version and a binary format.  A lot of the methods in AbstractRequest focus on serialization.. for example, AbstractRequest#getRequest deserializes, AbstractRequestResponse#sizeOf gets the size of the serialized data, etc.  You have to change the whole class hierarchy to modify this assumption.  Probably if you do that you would want to use a different name, such as Message, to represent the fact that the new class hierarchy you are creating has nothing to do with serialized data.  This might be worth doing.  It would be a very big change.

For #2, there are methods in NetworkClient such as leastLoadedNode that send the request to a node of the NetworkClient's own choosing.  This doesn't really work unless the NetworkClient is in charge of picking a version.  There also seems to be a bit of a layering violation here if the users of NC need to start concerning themselves with questions like has the peer they're talking to restarted or dropped its TCP session since we last learned about its version?

#1 might make sense as part of a general refactor of the RPC class hierarchies.  Getting rid of the Request class altogether might be a good idea since it invites confusion with ClientRequest.  I'm not sure I see a big win from #2 or a way to do it (but perhaps I am missing something).

ijuma commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -560,11 +569,22 @@ private void handleDisconnections(List<ClientResponse> responses, long now) {
+                log.debug("Completed connection to node {}.  Fetching API versions.", node);
+            } else {
+                // If the NetworkClient has not been configured to discover the versions of peers,
+                // go directly into the READY state.
```
Are the comments inside the if/else adding anything over what the code does? To me, they seem to say exactly the same thing as the code.

cmccabe replied:
not really, can remove

cmccabe committed:
NodeApiVersions: fix getUsableVersion, refactor toString

cmccabe committed:
ReplicaFetcherThread: some style fixes

cmccabe committed:
JoinGroupRequest: set rebalanceTimeout to -1 if the version is too old to send

cmccabe committed:
Some style changes

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/ControlledShutdownRequest.java:
```diff
@@ -16,33 +16,53 @@
 
-    private static final Schema CURRENT_SCHEMA = ProtoUtils.currentRequestSchema(ApiKeys.CONTROLLED_SHUTDOWN_KEY.id);
+    public static class Builder extends AbstractRequest.Builder<ControlledShutdownRequest> {
+        private int brokerId;
```
Seems this field could be final. Might be worth a pass over the other request types to locate other fields that could be final.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeGroupsRequest.java:
```diff
@@ -15,27 +15,52 @@
+            this.groupIds = groupIds;
+        }
+
+        public Builder setGroupIds(List<String> groupIds) {
```
This seems unneeded.

cmccabe committed:
NetworkClient: some style changes

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/JoinGroupRequest.java:
```diff
@@ -62,51 +60,103 @@ public ByteBuffer metadata() {
+        private String protocolType = null;
+        private List<ProtocolMetadata> groupProtocols = null;
+
+        public Builder(String groupId, int sessionTimeout, String memberId,
```
It would be nice to be consistent about whether a field should be included in the constructor or whether it should have a setter. I don't know that we have any cases that call for mutating a field that was initialized in the constructor, so I would prefer to just leave off the setters in that case and make the field final.

ijuma replied:
I did this for a bunch of builders in my PR.

cmccabe committed:
misc cleanups

* make nodeApiVersions keyed on String, not Integer
* some logging and whitespace cleanups
* NodeApiVersions#getUsableVersion takes an ApiKeys object, not a short

cmccabe committed:
NodeApiVersions: fix getUsableVersion

cmccabe committed:
ControllerChannelManager: fix bug where we were always setting the UpdateMetadataRequest version to 0

ijuma committed:
Various clean-ups in NodeApiVersions, Protocol and ApiKeys

* Better encapsulation in `Protocol`
* Higher-level code in NodeApiVersions
* Remove unneeded casts
* And more

ijuma committed:
Remove unused `ProtoUtils.parseResponse`

ijuma commented:
I submitted a PR with some clean-ups: https://github.com/cmccabe/kafka/pull/1

cmccabe commented:
@ijuma: Thanks for the cleanup pull request.  Looks good... I merged it into this branch.

ijuma committed:
Various builder and fetcher clean-ups

ijuma commented:
There are a number of failures in the system tests run. I started another one to double-check that it's not an environmental issue. In the meantime, I submitted another PR with clean-ups:

https://github.com/cmccabe/kafka/pull/2

I think we can merge after that (assuming the system tests pass) and do additional work in follow-up PRs.

ijuma committed:
Introduce overloaded `newClientRequest` that doesn’t take a callback

ijuma committed:
Remove unused code from `ControllerChannelManager` and `KafkaController`

cmccabe committed:
StreamsKafkaClient: rebase

cmccabe committed:
Centralize versionMismatch handling

cmccabe committed:
Fetcher.java: fix logging format string

cmccabe committed:
Fetcher: handle exception properly in handleListOffsetResponse

ijuma reviewed (approved):
Thanks for the changes, LGTM. We can address remaining issues in follow-ups. @cmccabe, please file a JIRA and link it from here when you can.

For the record, we had a known transient system test failure (`ZooKeeperSecurityUpgradeTest.test_zk_security_upgrade`), but it passed on rerun.

I fixed some generic unchecked warnings before merging to trunk.
